zebra-chain/src/primitives/redpallas/batch.rs

```diff
@@ -0,0 +1,265 @@
+    ///
+    /// As follows elliptic curve scalar multiplication convention,
+    /// scalar variables are lowercase and group point variables
+    /// are uppercase. This does not exactly match the RedDSA
+    /// notation in the [protocol specification §B.1][ps].
+    ///
+    /// [ps]: https://zips.z.cash/protocol/protocol.pdf#reddsabatchverify
+    #[allow(non_snake_case)]
+    pub fn verify<R: RngCore + CryptoRng>(self, mut rng: R) -> Result<(), Error> {
+        let n = self.signatures.len();
+
+        let mut VK_coeffs = Vec::with_capacity(n);
+        let mut VKs = Vec::with_capacity(n);
+        let mut R_coeffs = Vec::with_capacity(self.signatures.len());
+        let mut Rs = Vec::with_capacity(self.signatures.len());
+        let mut P_spendauth_coeff = pallas::Scalar::zero();
+        let mut P_binding_coeff = pallas::Scalar::zero();
+
+        for item in self.signatures.iter() {
+            let (s_bytes, r_bytes, c) = match item.inner {
+                Inner::SpendAuth { sig, c, .. } => (sig.s_bytes, sig.r_bytes, c),
+                Inner::Binding { sig, c, .. } => (sig.s_bytes, sig.r_bytes, c),
+            };
+
+            let s = {
+                // XXX-pallas: should not use CtOption here
+                let maybe_scalar = pallas::Scalar::from_bytes(&s_bytes);
+                if maybe_scalar.is_some().into() {
+                    maybe_scalar.unwrap()
+                } else {
+                    return Err(Error::InvalidSignature);
+                }
+            };
+
+            let R = {
+                // XXX-pallas: should not use CtOption here
+                // XXX-pallas: inconsistent ownership in from_bytes
+                let maybe_point = pallas::Affine::from_bytes(&r_bytes);
+                if maybe_point.is_some().into() {
+                    pallas::Point::from(maybe_point.unwrap())
+                } else {
+                    return Err(Error::InvalidSignature);
+                }
+            };
+
+            let VK = match item.inner {
+                Inner::SpendAuth { vk_bytes, .. } => {
+                    VerificationKey::<SpendAuth>::try_from(vk_bytes.bytes)?.point
+                }
+                Inner::Binding { vk_bytes, .. } => {
+                    VerificationKey::<Binding>::try_from(vk_bytes.bytes)?.point
+                }
+            };
+
+            let z = pallas::Scalar::from_raw(gen_128_bits(&mut rng));
+
+            let P_coeff = z * s;
+            match item.inner {
+                Inner::SpendAuth { .. } => {
+                    P_spendauth_coeff -= P_coeff;
+                }
+                Inner::Binding { .. } => {
+                    P_binding_coeff -= P_coeff;
+                }
+            };
+
+            R_coeffs.push(z);
+            Rs.push(R);
+
+            VK_coeffs.push(pallas::Scalar::zero() + (z * c));
```

Comment:
Is `pallas::Scalar::zero() +` needed? Seems to work without it

Reply:
Let me check the formulas, this may have been to match very closely

Reply:
I think you are right

Reply:
Done and tested in 580f6f95